pip/pip-269.md

```diff
@@ -0,0 +1,239 @@
+| time | **Reset cursor** | **Consumer reconnect** |
+| --- | --- | --- |
+| 1 | disconnect all consumers |
+| 2 | | remove consumer |
+| 3 | | add consumer_1 |
+| 4 | | rewind cursor to `3:0` |
+| 5 | | add consumer_2 |
+| 6 | | add consumer_2 to recently joined consumers, `max read position` is `3:0`<sup>[4]</sup> |
+| 7 | set read position to `1:0` |
+
+In this scenario(referred to as **Scenario-3**), the `consumer_2` will be stuck by the mechanism Recently Joined Consumers of Key_Shared mode<sup>[4]</sup>. It's a simple question, but it's more complicated than that:
+- If there has an in-flight reading.
+- If there is more than one **Occasional Read Position Change** action running.
+- Both above.
+
+---
+
+In the Multi Consumer Dispatcher, there has a mechanism that tries to fix the issue caused by the action `cursor.rewind` and **Sequentially Read** executed concurrently(referred to as Rewind After In-flight Reading). It works like this: if there has an in-flight reading when calling `cursor.rewind`, just do rewind after the in-flight reading is finished.` but this mechanism has the following drawbacks:
+- Makes a new issue lead to a new issue<sup>[5]</sup>.
+- Only applicable to `cursor.rewind`.
+- Only applicable to Multi Consumer Dispatcher.
+- If there is more than one in-flight reading(one Sequentially Read and some Message Replay Read might exist simultaneously), this mechanism can not work.
+
+# Goals
+- For the **Scenario-1** and **Scenario-2**(see Motivation), discard the in-flight reading(referred to as **Goal-1**).
+- For the **Scenario-3**(see Motivation), we should do these two things:
+    - Provide a way to guarantee the atomicity modification of both attributes of Cursor and attributes of Dispatcher(referred to as **Goal-2**).
+    - Prevent more than one **Occasional Read Position Change** executed at the same time(Current Proposal does not consider this optimization, there will be some bug fixes after this PIP)
+- Revert https://github.com/apache/pulsar/issues/6403
+- Remove the mechanism Rewind After In-flight Reading<sup>[5]</sup>
+
+# High-Level Design
+
+**For the Goal-1**: Cursor Epoch
+- Definition of a counter to record how many times the `read position` of the cursor has been changed by the **Occasional Read Position Change**, we call it Cursor Epoch. **Occasional Read Position Change** executed once, the Cursor Epoch increase once.
+- Every **Sequentially Read** or **Message Replay Read** carries the current Cursor Epoch.
+- Discard the reading and prevent the `read position` change if the epoch carried by the reading is smaller than the current Cursor Epoch when the reading complete
+
+**For the Goal-2**: Two-phase of Cursor Epoch increment
```

Comment:
Lost you here.
Why do you need to use same epoch.
I wonder why not a lock perhaps for cursor and dispatcher.
I guess I need to dive more perhaps.

Reply:
> Lost you here.

I have rewritten the doc to make it easier to understand.

> Why do you need to use same epoch.

I could not fully understand this.

> I wonder why not a lock perhaps for cursor and dispatcher.
> I guess I need to dive more perhaps.

- The lock is better than `synchronized` because it is helpful for async logic.
- A lock and a variable is better than only a lock because they can help correct the task already in the queue of lock. 
- Two locks (one for Cursor and one for Dispatcher) can make the lock range smaller. For example:   
  - lock(Cursor), set `resetting` to `true`, release lock(Cursor). Other actions will be rejected by `resetting` in progress(fast fail).
  - lock(Dispatcher), do itself progress.

Since there is more than one component called **Cursor Position Reset**(see the section **Cursor Position Reset**, I have rewritten it), we can't make the lock too wide, it would increase the risk of deadlock.